app/queries/servers/thread.ts

```diff
@@ -107,15 +109,18 @@ export const prepareThreadsFromReceivedPosts = async (operator: ServerDataOperat
             toUpdate[post.root_id] = Math.max(toUpdate[post.root_id] || 0, post.create_at, post.update_at, post.delete_at);
         }
     });
+
     if (threads.length) {
-        const threadModels = await operator.handleThreads({threads, prepareRecordsOnly: true});
+        const threadModels = await operator.handleThreads({threads, prepareRecordsOnly: true}) as ThreadModel[];
+        processedThreads = new Set<string>(threadModels.map((t) => t.id));
         models.push(...threadModels);
     }
+
     const toUpdateKeys = Object.keys(toUpdate);
     if (toUpdateKeys.length) {
         const toUpdateThreads = await Promise.all(toUpdateKeys.map((key) => getThreadById(operator.database, key)));
         for (const thread of toUpdateThreads) {
-            if (thread) {
+            if (thread && !processedThreads?.has(thread.id)) {
```

Comment:
Wouldn't this be O(n^2); for each thread we loop through processThreads [Set.has()](https://262.ecma-international.org/6.0/#sec-set.prototype.has) seems to loop through the list rather than a direct look up.

Might be better to do `processThreads = {}` and then doing something like 

```
for( const tm of threadModels ) processThreads[`${tm.id}`] = tm.id
```

And then you can just check directly via `if (thread && !processThreads[thread.id])`

Reply:
@shazm a set is direct access O(1)

Reply:
See https://stackoverflow.com/questions/33611509/es6-map-and-set-complexity-v8-implementation